library/kani_core/src/arbitrary.rs

```diff
@@ -169,6 +180,171 @@ macro_rules! generate_arbitrary {
         arbitrary_tuple!(A, B, C, D, E, F, G, H, I, J);
         arbitrary_tuple!(A, B, C, D, E, F, G, H, I, J, K);
         arbitrary_tuple!(A, B, C, D, E, F, G, H, I, J, K, L);
+
+        /// Enumeration with the cases currently covered by the pointer generator.
+        #[derive(Copy, Clone, Debug, PartialEq, Eq, kani::Arbitrary)]
+        pub enum AllocationStatus {
+            /// Dangling pointers
+            Dangling,
+            /// Pointer to dead object
+            DeadObj,
+            /// Null pointers
+            Null,
+            /// In bounds pointer (it may be unaligned)
+            InBounds,
+            /// Out of bounds
+            OutBounds,
+        }
+
+        /// Pointer generator that can be used to generate an arbitrary pointer.
+        ///
+        /// This generator allows users to build pointers with different safety properties.
+        /// It contains an internal buffer that it uses to generate `InBounds` and `OutBounds` pointers.
+        /// In those cases, the pointers will have the same provenance as the generator, and the same lifetime.
+        ///
+        /// For example:
+        /// ```ignore
+        ///     let generator = PointerGenerator<char, 5>::new();
+        ///     let arbitrary = generator.generate_ptr();
+        ///     kani::assume(arbitrary.status == AllocationStatus::InBounds);
+        ///     // Pointer may be unaligned but it should be in-bounds.
+        ///     unsafe { arbitrary.ptr.write_unaligned(kani::any() }
+        /// ```
+        ///
+        /// Use the same generator if you want to handle cases where 2 or more pointers may overlap. E.g.:
+        /// ```ignore
+        ///     let generator = PointerGenerator<char, 5>::new();
+        ///     let arbitrary = generator.generate_ptr();
+        ///     kani::assume(arbitrary.status == AllocationStatus::InBounds);
+        ///     let ptr = arbitrary.ptr;
+        ///     kani::cover!(arbitrary.ptr == generator.generate_ptr());
+        ///     kani::cover!(arbitrary.ptr != generator.generate_ptr());
+        /// ```
+        ///
+        /// Note: This code is different than generating a pointer with any address. I.e.:
+        /// ```ignore
+        ///     // This pointer represents any address.
+        ///     let ptr = kani::any::<usize>() as *const u8;
+        ///     // Which is different from:
+        ///     let generator = PointerGenerator<u8, 5>::new();
+        ///     let ptr = generator.generate_ptr().ptr;
```

Comment:
You say it is "different," but could you please elaborate in what way this is different?